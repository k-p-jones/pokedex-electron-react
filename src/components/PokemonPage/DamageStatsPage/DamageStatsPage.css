.damage-stats-page-type-header::first-letter {
  text-transform: uppercase;
}
