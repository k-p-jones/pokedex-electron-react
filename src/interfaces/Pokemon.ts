interface BasicObject {
  name: string;
  url: string;
}

export default interface Pokemon {
  id?: number;
  name: string;
  speciesURL: string;
  dataURL: string;
  data: {
    types: {
      name: string,
      url: string,
      data: {
        double_damage_from: BasicObject[];
        double_damage_to: BasicObject[];
        half_damage_from: BasicObject[];
        half_damage_to: BasicObject[];
        no_damage_from: BasicObject[];
        no_damage_to: BasicObject[];
      };
    }[];
    stats: {
      base_stat: number;
      effort: number;
      stat: BasicObject;
    }[];
  };
  color?: string | 'white';
  imageURL: string;
  dataFetched: boolean;
  typeDataFetched: boolean;
}
