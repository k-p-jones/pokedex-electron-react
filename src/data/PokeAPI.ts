import axios from 'axios';
import BaseApiResponse from '../interfaces/BaseApiResponse';
import SpeciesApiResponse from '../interfaces/SpeciesApiResponse';
import Pokemon from '../interfaces/Pokemon';
import COLORS from '../consts/colors';
import PokemonBaseType from '../interfaces/PokemonBaseType';

class PokeAPI {
  fetchBasePokemon = () => {
    const pokemon: Array<Pokemon> = [];

    return new Promise<Pokemon[]>((resolve, reject) => {
      try {
        axios.get('https://pokeapi.co/api/v2/pokemon-species?limit=150').then((response) => {
          response.data.results.forEach((data: BaseApiResponse, index: number) => {
            const obj: Pokemon = {
              id: index + 1,
              name: data.name,
              speciesURL: data.url,
              dataURL: `https://pokeapi.co/api/v2/pokemon/${index + 1}`,
              imageURL: `https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/other/official-artwork/${index + 1}.png`,
              evolutionChainURL: '',
              data: {
                types: [],
                stats: []
              },
              dataFetched: false,
              typeDataFetched: false,
              evolutionChainFetched: false
            }
            pokemon.push(obj);
          });
        }).then(async () => {
          await Promise.all(
            pokemon.map(async (p: Pokemon) => {
              const response = await axios.get(p.speciesURL);
              const data: SpeciesApiResponse = response.data;
              p.color = COLORS[data.color.name];
              p.evolutionChainURL = data.evolution_chain.url;
            })
          );
          return resolve(pokemon);
        })
      } catch (error) {
        reject(error);
      }
    });
  }

  fetchPokemonData = (pokemon: Pokemon) => {
    return new Promise<void>(async (resolve) => {
      if (pokemon.dataFetched) { resolve() }
      const response = await axios.get(pokemon.dataURL);
      const types = response.data.types.map((t: PokemonBaseType) => {
        return t.type;
      });
      pokemon.data.types = types;
      pokemon.data.stats = response.data.stats;
      pokemon.dataFetched = true;
      resolve();
    })
  }

  fetchTypeData = (pokemon: Pokemon) => {
    return new Promise<void>(async (resolve) => {
      if (pokemon.typeDataFetched) { return resolve() }
      await Promise.all(
        pokemon.data.types.map(async (type, index) => {
          const response = await axios.get(type.url);
          pokemon.data.types[index].data = response.data.damage_relations;
        })
      );
      pokemon.typeDataFetched = true;
      resolve();
    });
  }

  fetchEvolutionChain = (pokemon: Pokemon) => {
    return new Promise<void>(async (resolve) => {
      if (pokemon.evolutionChainFetched) { return resolve() }
      const response = await axios.get(pokemon.evolutionChainURL)
      pokemon.data.evolutionChain = response.data.chain;
      pokemon.evolutionChainFetched = true;
      resolve();
    });
  }
}

export default new PokeAPI();
