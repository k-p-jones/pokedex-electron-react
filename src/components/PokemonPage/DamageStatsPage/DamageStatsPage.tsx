import React, { useEffect, useState } from 'react';
import { Loader, Segment } from 'semantic-ui-react';
import PokeAPI from '../../../data/PokeAPI';
import BasicObject from '../../../interfaces/BasicObject';
import Pokemon from '../../../interfaces/Pokemon';
import './DamageStatsPage.css';

interface Props {
  pokemon: Pokemon;
}

const DamageStatsPage: React.FC<Props> = ({ pokemon }) => {
  const [isLoading, setIsLoading] = useState<boolean>(true);

  useEffect(() => {
    PokeAPI.fetchTypeData(pokemon).then(() => setIsLoading(false));
  }, [pokemon])

  const strongAttacking = (data: BasicObject[]) => {
    const types = data.map((t) => <li key={`type-${t.name}-2x-damage-given`}>{t.name}</li>)

    // Normal type has no 2x damage modifiers
    if (types.length === 0) { return null }

    return (
      <React.Fragment>
        <h3>Strong Attacking (2x Damage)</h3>
        <ul>{types}</ul>
      </React.Fragment>
    )
  }

  const weakAttacking = (halfDamage: BasicObject[], noDamage: BasicObject[]) => {
    const hd = halfDamage.map(t => <li key={`type-${t.name}-half-damage-given`}>{t.name}</li>);
    const nd = noDamage.map(t => <li key={`type-${t.name}-no-damage-given`}>{t.name}</li>);

    let halfDamageDisplay;
    let noDamageDisplay;

    if (hd.length > 0) {
      halfDamageDisplay = (
        <React.Fragment>
          <h3>Weak Attacking (1/2 Damage)</h3>
          <ul>{hd}</ul>
        </React.Fragment>
      );
    }

    if (nd.length > 0) {
      noDamageDisplay = (
        <React.Fragment>
          <h3>Weak Attacking (No Damage)</h3>
          <ul>{nd}</ul>
        </React.Fragment>
      );
    }

    return (
      <React.Fragment>
        {halfDamageDisplay}
        {noDamageDisplay}
      </React.Fragment>
    )
  }

  const strongDefending = (halfDamage: BasicObject[], noDamage: BasicObject[]) => {
    const hd = halfDamage.map(t => <li key={`type-${t.name}-half-damage`}>{t.name}</li>);
    const nd = noDamage.map(t => <li key={`type-${t.name}-no-damage`}>{t.name}</li>);

    let halfDamageDisplay;
    let noDamageDisplay;

    if (hd.length > 0) {
      halfDamageDisplay = (
        <React.Fragment>
          <h3>Strong Defending (1/2 Damage)</h3>
          <ul>{hd}</ul>
        </React.Fragment>
      );
    }

    if (nd.length > 0) {
      noDamageDisplay = (
        <React.Fragment>
          <h3>Strong Defending (Immune)</h3>
          <ul>{nd}</ul>
        </React.Fragment>
      );
    }

    return (
      <React.Fragment>
        {halfDamageDisplay}
        {noDamageDisplay}
      </React.Fragment>
    )
  }

  const weakDefending = (data: BasicObject[]) => {
    const types = data.map((t) => <li key={`type-${t.name}-2x-damage-recieved`}>{t.name}</li>)

    return (
      <React.Fragment>
        <h3>Weak Defending (2x Damage)</h3>
        <ul>{types}</ul>
      </React.Fragment>
    )
  }

  const damageStats = () => (
    pokemon.data.types.map((t) => {
      return (
        <Segment key={`type-${t.name}-data`}>
          <h1 className="damage-stats-page-type-header">{t.name}</h1>
          {strongAttacking(t.data.double_damage_to)}
          {strongDefending(t.data.half_damage_from, t.data.no_damage_from)}
          {weakAttacking(t.data.half_damage_to, t.data.no_damage_to)}
          {weakDefending(t.data.double_damage_from)}
        </Segment>
      )
    })
  )

  if (isLoading) {
    return <Loader />
  }

  return (<React.Fragment>{damageStats()}</React.Fragment>)
}

export default DamageStatsPage;
