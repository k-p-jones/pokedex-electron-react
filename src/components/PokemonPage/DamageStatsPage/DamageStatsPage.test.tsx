import React from 'react';
import { render, screen, waitFor } from '@testing-library/react';
import DamageStatsPage from './DamageStatsPage';
import PokeAPI from '../../../data/PokeAPI';
import Pokemon from '../../../interfaces/Pokemon';

afterEach(() => {
  jest.resetAllMocks();
});

const pokemon = {
  id: 1,
  name: 'bulbasaur',
  speciesURL: 'https://pokeapi.co/api/v2/pokemon-species/1',
  dataURL: 'https://pokeapi.co/api/v2/pokemon/1',
  imageURL: 'https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/other/official-artwork/1.png',
  data: {
    types: [
      {
        name: 'poison',
        url: 'path/to/pokemon',
        data: {
          double_damage_from: [],
          double_damage_to: [],
          half_damage_from: [],
          half_damage_to: [],
          no_damage_from: [],
          no_damage_to: []
        }
      },
      {
        name: 'grass',
        url: 'path/to/pokemon',
        data: {
          double_damage_from: [],
          double_damage_to: [],
          half_damage_from: [],
          half_damage_to: [],
          no_damage_from: [],
          no_damage_to: []
        }
      }
    ],
    stats: []
  },
  color: 'green',
  evolutionChainURL: 'https://pokeapi.co/api/v2/evolution-chain/1',
  dataFetched: false,
  typeDataFetched: false,
  evolutionChainFetched: false
}

const poison = {
  double_damage_from: [
    {
      name: "ground",
      url: "https://pokeapi.co/api/v2/type/5/"
    },
    {
      name: "psychic",
      url: "https://pokeapi.co/api/v2/type/14/"
    }
  ],
  double_damage_to: [
    {
      name: "grass",
      url: "https://pokeapi.co/api/v2/type/12/"
    },
    {
      name: "fairy",
      url: "https://pokeapi.co/api/v2/type/18/"
    }
  ],
  half_damage_from: [
    {
      name: "fighting",
      url: "https://pokeapi.co/api/v2/type/2/"
    },
    {
      name: "poison",
      url: "https://pokeapi.co/api/v2/type/4/"
    },
    {
      name: "bug",
      url: "https://pokeapi.co/api/v2/type/7/"
    },
    {
      name: "grass",
      url: "https://pokeapi.co/api/v2/type/12/"
    },
    {
      name: "fairy",
      url: "https://pokeapi.co/api/v2/type/18/"
    }
  ],
  half_damage_to: [
    {
      name: "poison",
      url: "https://pokeapi.co/api/v2/type/4/"
    },
    {
      name: "ground",
      url: "https://pokeapi.co/api/v2/type/5/"
    },
    {
      name: "rock",
      url: "https://pokeapi.co/api/v2/type/6/"
    },
    {
      name: "ghost",
      url: "https://pokeapi.co/api/v2/type/8/"
    }
  ],
  no_damage_from: [],
  no_damage_to: [
    {
      name: "steel",
      url: "https://pokeapi.co/api/v2/type/9/"
    }
  ]
}

const grass = {
  double_damage_from: [
    {
      name: "flying",
      url: "https://pokeapi.co/api/v2/type/3/"
    },
    {
      name: "poison",
      url: "https://pokeapi.co/api/v2/type/4/"
    },
    {
      name: "bug",
      url: "https://pokeapi.co/api/v2/type/7/"
    },
    {
      name: "fire",
      url: "https://pokeapi.co/api/v2/type/10/"
    },
    {
      name: "ice",
      url: "https://pokeapi.co/api/v2/type/15/"
    }
  ],
  double_damage_to: [
    {
      name: "ground",
      url: "https://pokeapi.co/api/v2/type/5/"
    },
    {
      name: "rock",
      url: "https://pokeapi.co/api/v2/type/6/"
    },
    {
      name: "water",
      url: "https://pokeapi.co/api/v2/type/11/"
    }
  ],
  half_damage_from: [
    {
      name: "ground",
      url: "https://pokeapi.co/api/v2/type/5/"
    },
    {
      name: "water",
      url: "https://pokeapi.co/api/v2/type/11/"
    },
    {
      name: "grass",
      url: "https://pokeapi.co/api/v2/type/12/"
    },
    {
      name: "electric",
      url: "https://pokeapi.co/api/v2/type/13/"
    }
  ],
  half_damage_to: [
    {
      name: "flying",
      url: "https://pokeapi.co/api/v2/type/3/"
    },
    {
      name: "poison",
      url: "https://pokeapi.co/api/v2/type/4/"
    },
    {
      name: "bug",
      url: "https://pokeapi.co/api/v2/type/7/"
    },
    {
      name: "steel",
      url: "https://pokeapi.co/api/v2/type/9/"
    },
    {
      name: "fire",
      url: "https://pokeapi.co/api/v2/type/10/"
    },
    {
      name: "grass",
      url: "https://pokeapi.co/api/v2/type/12/"
    },
    {
      name: "dragon",
      url: "https://pokeapi.co/api/v2/type/16/"
    }
  ],
  no_damage_from: [],
  no_damage_to: []
}

const fetchTypeData = (p: Pokemon) => {
  const damageData = [poison, grass];
  return new Promise<void>((resolve) => {
    p.data.types.forEach((t: { data: object }, index: number) => t.data = damageData[index]);
    resolve();
  });
}

beforeEach(() => {
  PokeAPI.fetchTypeData = jest.fn((p) => fetchTypeData(p));
});

test('it displays the content', async () => {
  const { container } = render(<DamageStatsPage pokemon={pokemon} />);
  // Test the loading content displays
  const spinnerText = screen.getByText('Loading...');
  const image = document.querySelector('img') as HTMLImageElement;
  expect(spinnerText).toBeInTheDocument();
  expect(image.src).toContain('pokeball.svg');

  await waitFor(() => {
    // Test the stats display
    const heading = screen.getAllByText('Strong Attacking (2x Damage)')
    expect(heading.length).toEqual(2);
    expect(container).toMatchSnapshot();
  })
});
