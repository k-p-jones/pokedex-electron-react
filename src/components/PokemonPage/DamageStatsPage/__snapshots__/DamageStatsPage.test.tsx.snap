// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it displays the content 1`] = `
<div>
  <div
    class="ui segment"
  >
    <h1
      class="damage-stats-page-type-header"
    >
      poison
    </h1>
    <h3>
      Strong Attacking (2x Damage)
    </h3>
    <ul>
      <li>
        grass
      </li>
      <li>
        fairy
      </li>
    </ul>
    <h3>
      Strong Defending (1/2 Damage)
    </h3>
    <ul>
      <li>
        fighting
      </li>
      <li>
        poison
      </li>
      <li>
        bug
      </li>
      <li>
        grass
      </li>
      <li>
        fairy
      </li>
    </ul>
    <h3>
      Weak Attacking (1/2 Damage)
    </h3>
    <ul>
      <li>
        poison
      </li>
      <li>
        ground
      </li>
      <li>
        rock
      </li>
      <li>
        ghost
      </li>
    </ul>
    <h3>
      Weak Attacking (No Damage)
    </h3>
    <ul>
      <li>
        steel
      </li>
    </ul>
    <h3>
      Weak Defending (2x Damage)
    </h3>
    <ul>
      <li>
        ground
      </li>
      <li>
        psychic
      </li>
    </ul>
  </div>
  <div
    class="ui segment"
  >
    <h1
      class="damage-stats-page-type-header"
    >
      grass
    </h1>
    <h3>
      Strong Attacking (2x Damage)
    </h3>
    <ul>
      <li>
        ground
      </li>
      <li>
        rock
      </li>
      <li>
        water
      </li>
    </ul>
    <h3>
      Strong Defending (1/2 Damage)
    </h3>
    <ul>
      <li>
        ground
      </li>
      <li>
        water
      </li>
      <li>
        grass
      </li>
      <li>
        electric
      </li>
    </ul>
    <h3>
      Weak Attacking (1/2 Damage)
    </h3>
    <ul>
      <li>
        flying
      </li>
      <li>
        poison
      </li>
      <li>
        bug
      </li>
      <li>
        steel
      </li>
      <li>
        fire
      </li>
      <li>
        grass
      </li>
      <li>
        dragon
      </li>
    </ul>
    <h3>
      Weak Defending (2x Damage)
    </h3>
    <ul>
      <li>
        flying
      </li>
      <li>
        poison
      </li>
      <li>
        bug
      </li>
      <li>
        fire
      </li>
      <li>
        ice
      </li>
    </ul>
  </div>
</div>
`;
